enterprise/frontend/src/metabase-enterprise/transforms/components/IncrementalTransform/KeysetColumnSelect/KeysetColumnSelect.tsx

```diff
@@ -0,0 +1,94 @@
+import { useMemo } from "react";
+
+import { FormSelect } from "metabase/forms";
+import * as Lib from "metabase-lib";
+
+type KeysetColumnSelectProps = {
+  name: string;
+  label: string;
+  placeholder: string;
+  description: string;
+  query: Lib.Query | null;
+  disabled?: boolean;
+};
+
+export function KeysetColumnSelect({
+  name,
+  label,
+  placeholder,
+  description,
+  query,
+  disabled,
+}: KeysetColumnSelectProps) {
+  const columnOptions = useMemo((): Array<{ value: string; label: string }> => {
+    if (!query) {
+      return [];
+    }
+
+    try {
+      // Use -1 to get the last stage
+      const stageIndex = -1;
+
+      // Get returned columns and filterable columns
+      const returnedColumns = Lib.returnedColumns(query, stageIndex);
+      const filterableColumns = Lib.filterableColumns(query, stageIndex);
+
+      // Create a set of filterable column identifiers using display info
+      const filterableIdentifiers = new Set(
+        filterableColumns.map((col) => {
+          const info = Lib.displayInfo(query, stageIndex, col);
+          return `${info.table?.name ?? "no-table"}:${info.name}`;
+        }),
+      );
+
+      // Filter returned columns to only those that are also filterable and numeric
+      const numericFilterableColumns = returnedColumns.filter((column) => {
+        const info = Lib.displayInfo(query, stageIndex, column);
+        const identifier = `${info.table?.name ?? "no-table"}:${info.name}`;
```

Comment:
Same - do not rely on the source being a table. There can be a transform with a Model as Data source + another Model used for join, and you would get duplicate keys with the current implementation